packages/vue-app/template/plugins.js

```diff
@@ -0,0 +1,124 @@
+import Vue from 'vue'
+<% if (features.hookablePlugins) {
+%>import Hookable from 'hable'
+<%
+}
+
+if (features.parallelPlugins) {
+%>import { interopDefault } from './utils'
+
+<%
+  for (const pluginMode of ['all', 'server', 'client']) {
+    const modePlugins = plugins.filter(({ mode }) => mode === pluginMode)
+    if (!modePlugins.length) {
+      continue
+    }
+
+    if (pluginMode === 'all') {
+      %>const pluginImports = [
+<%
+    } else {
+      %>if (process.<%= pluginMode %>) {
+  pluginImports.concat([
+<%
+    }
+
+    modePlugins.forEach((plugin, index, array) => {
+      const sep = index === (array.length - 1) ? '' : ','
+      const space = '  '.repeat(pluginMode === 'all' ? 1 : 2)
+
+      %><%= space %>import('<%= plugin.name %>')<%= sep %> // Source: <%= relativeToBuild(plugin.src) %> (mode: '<%= plugin.mode %>')
+<%
+    })
+
+    if (pluginMode === 'all') {
+      %>]
+<%
+    } else {
+      %>  ])
+}
+<%
+    }
+  }
+} else { %>
+<%= isTest ? '/* eslint-disable camelcase */' : '' %>
+<% plugins.forEach((plugin) => { %>import <%= plugin.name %> from '<%= plugin.name %>' // Source: <%= relativeToBuild(plugin.src) %> (mode: '<%= plugin.mode %>')
+<% }) %>
+<%= isTest ? '/* eslint-enable camelcase */' : '' %>
+<% } %>
+
+export default async function executePlugins(app<%= store ? ', store' : '' %>) {
+  <% if (features.hookablePlugins) { %>
+  const hooks = new Hookable()
+  <% } %>
+
+  function inject (key, value) {
+    if (!key || value === undefined) {
+      throw new Error(`inject(key, value) has no ${key ? 'value' : 'key'} provided`)
+    }
+
+    key = `$${key}`
+    // Add into app
+    app[key] = value
+    <% if (store) { %>
+    // Add into store
+    store[key] = app[key]
+    <% } %>
+
+    if (Vue.prototype.hasOwnProperty(key)) {
+      return
+    }
+
+    Object.defineProperty(Vue.prototype, key, {
+      get () {
+        return this.$root.$options[key]
+      }
+    })
+  }
+
+  <% if (features.parallelPlugins) { %>
+  async function loadPlugin (plugin) {
+    if (!plugin || typeof plugin !== 'function') {
+      return
+    }
+
+    let injectPromises = []
+    const asyncInject = (key, value) => {
```

Comment:
This is not really a nice solution and actually the `injected:<key>` hook doesnt work yet for `parallelPlugins`. It would be much easier if we could change the `inject` function itself to be async, but not sure if thats a BC or not

Reply:
Or maybe we should support synchronous hooks calling